mimesniff/mime-type-sniffing/resources/generate_scriptable_mime_type_tests.py

```diff
@@ -0,0 +1,145 @@
+  PatternTuple( True, "3C 54 49 54 4C 45 TT", "FF DF DF DF DF DF FF", True, "text/html" ),
+  PatternTuple( True, "3C 42 TT", "FF DF FF", True, "text/html" ),
+  PatternTuple( True, "3C 42 4F 44 59 TT", "FF DF DF DF DF FF", True, "text/html" ),
+  PatternTuple( True, "3C 42 52 TT", "FF DF DF FF", True, "text/html" ),
+  PatternTuple( True, "3C 50 TT", "FF DF FF", True, "text/html" ),
+  PatternTuple( True, "3C 21 2D 2D TT", "FF FF FF FF FF", True, "text/html" ),
+  PatternTuple( True, "3C 3F 78 6D 6C", "FF FF FF FF FF", True, "text/xml" ),
+  PatternTuple( True, "25 50 44 46 2D", "FF FF FF FF FF", False, "application/pdf" ),
+  PatternTuple( False, "25 21 50 53 2D 41 64 6F 62 65 2D", "FF FF FF FF FF FF FF FF FF FF FF", False, "application/postscript" ),
+  PatternTuple( False, "FE FF 00 00", "FF FF 00 00", False, "text/plain" ),
+  PatternTuple( False, "FF FE 00 00", "FF FF 00 00", False, "text/plain" ),
+  PatternTuple( False, "EF BB BF 00", "FF FF FF 00", False, "text/plain" ),
+]
+
+#####
+
+def byte_pattern_is_binary(byte_pattern: str) -> bool:
+  return any([ dbd.hex().upper() in byte_pattern for dbd in BINARY_DATA_BYTES ])
+
+def get_description(byte_pattern: str, content_type: str, nosniff: bool) -> str:
+  pattern_bytes: bytes = bytes.fromhex(payload)
+
+  description: str
+  if byte_pattern_is_binary(byte_pattern):
+    description = repr(pattern_bytes)
+  else:
+    description = repr(pattern_bytes.decode())
+
+  if content_type is None:
+    description += " with no Content-Type"
+  elif content_type == "":
+    description += " with empty Content-Type"
+  else:
+    description += f" with '{content_type}' Content-Type"
+
+  if nosniff:
+    description += " and nosniff"
+
+  return description
+
+def get_safe_expected_mime_type(byte_pattern: str) -> str:
+  return "application/octet-stream" if byte_pattern_is_binary(byte_pattern) else "text/plain"
+
+#####
+
+tests = []
+
+for content_type in unknown_content_types:
+  for pattern_tuple in pattern_tuples:
+    ignored_byte_prefixes = [ None ]
+
+    if pattern_tuple.ignore_leading_whitespace:
+      ignored_byte_prefixes.append(" ".join([ ws.hex().upper() for ws in WHITESPACE_BYTES ]))
+
+      for whitespace_byte in WHITESPACE_BYTES:
+        ignored_byte_prefixes.append(" ".join([ whitespace_byte.hex().upper() ] * len(WHITESPACE_BYTES)))
+
+    for ignored_byte_prefix in ignored_byte_prefixes:
+      for tag_terminating_byte in ( None, ) + TAG_TERMINATING_BYTES:
+        for nosniff in (False, True):
+          # Don't attempt to generate a separate test expanding 'TT' if it's not even in the byte pattern.
+          if tag_terminating_byte is not None and "TT" not in pattern_tuple.byte_pattern:
+            continue
+
+          # Indicates that this is a derived pattern that should *NOT* match the expected MIME type.
+          pattern_does_not_match: bool = (tag_terminating_byte is None and "TT" in pattern_tuple.byte_pattern)
```

Comment:
Rather than doing two "TT" searches here, do a single search first?  Seems clearer than redundant searches, even if it's a bit more code.

Reply:
This mess is a side effect of the mess in the spec, which I'd prefer to simply eliminate entirely at a later point. I don't know if it's worth fixing right now?